spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,331 @@
+
+#### What k8s-infra does
+k8s-infra uses the same `KnownTypeReference` type mentioned above for ownership references too.
+There are two patterns for ownership in k8s-infra today.
+
+One pattern is used for ResourceGroup, where 
+[top level resources have a link to the resource group they are in](https://github.com/Azure/k8s-infra/blob/14105b1cb3f6967cd086c9f8f75fb16bb85d6318/apis/microsoft.compute/v1/virtualmachinescaleset_types.go#L323).
+
+```go
+type VirtualMachineScaleSetSpec struct {
+    // +kubebuilder:validation:Required
+    ResourceGroupRef *azcorev1.KnownTypeReference `json:"resourceGroupRef" group:"microsoft.resources.infra.azure.com" kind:"ResourceGroup"`
+    ...
+}
+```
+
+The other pattern is where the owning resource has links to the dependent resources it is expecting to have:
+
+```go
+type RouteTableSpecProperties struct {
+    DisableBGPRoutePropagation bool                          `json:"disableBgpRoutePropagation,omitempty"`
+    RouteRefs                  []azcorev1.KnownTypeReference `json:"routeRefs,omitempty" group:"microsoft.network.infra.azure.com" kind:"Route" owned:"true"`
+}
+```
+
+If the dependent resources aren't there, the status of the owning resource reflects an error explaining that.
+
+## Proposal
+
+### How to represent references
+We construct a new type which is our reference holder. I am calling this approach `AugmentedKnownTypeReference` (real type name TBD), because
+it's similar to the current approach that `k8s-infra` is taking but with a few changes.
+
+```go
+type AugmentedKnownTypeReference struct {
+    Kubernetes KubernetesReference
+    Azure      AzureReference
+}
+
+type KubernetesReference struct {
+    // The "Group" of the owning resources GVK. There may be an annotation on the property 
+    // that specifies Group, in which case this is optional (if supplied it must match the group
+    // on the annotation). For resources which do not have a group annotation, this is required.
+    // TODO: It's possibly going to be confusing to customers when they have to specify this and when they don't...
+    Group string
+
+    // The "Kind" of the owning resources GVK. There may be an annotation on the property 
+    // that specifies Kind, in which case this is optional (if supplied it must match the kind
+    // on the annotation). For resources which do not have a kind annotation, this is required.
+    // TODO: It's possibly going to be confusing to customers when they have to specify this and when they don't... 
+    Kind string
+
+    // This is the name of the kubernetes resource to reference. 
+    // References across namespaces are not supported.
+    Name string
+
+    // Note that ownership across namespaces in Kubernetes is not allowed, but technically resource
+    // references are. There are RBAC considerations here though so probably easier to just start by 
+    // disallowing cross-namespace references
+}
+
+type AzureReference struct {
```

Comment:
Did you consider just making this an Azure resourceID string? I think you are going to run into problems with only having Sub, Group, and Name.

Reply:
Sub, Group and Name (with how I possibly poorly defined name) is == to fully qualified ARM ID.
but yes we could just make this FQArmId and it might make our life easier.

The main advantage of this was just to have defaulting where if you're in a single sub you don't have to keep repeating the sub in all of your YAML (which presumably would get annoying if you wanted to deploy to a different sub, although maybe Kustomize or something can help with that and we don't want to solve it here?)

I suppose the other argument for "just make this the full ARM ID" is that if they really want nice transportable YAMLs they should be using the Kubernetes flavored references, which will all translate quite nicely between namespaces

Reply:
I'm certainly open to changing this - there are some advantages for "similarity to ARM templates" too

Reply:
Related, but the authoring experience could become disjointed if one is often working with existing infrastructure via Azure'isms when they are used to K8s'isms. Just something to consider when bridging the gap.

Reply:
I've added a whole section on this (`Open Design Question - shape of Azure References`).

I've asked my team to weigh in as well.